### ISSUE TITLE
[Perf] `vsa_blackwell` BSA forward plateaus at ~900–960 TFLOP/s on GB200 (SM100) — ~60–65% of an FA4-lineage CuTe-DSL BSA on identical block masks

### ISSUE THREAD

**Issue by SolitaryThinker** [flashinfer-ai/flashinfer#4554]
We benchmarked `BlockSparseAttentionWrapper(backend="vsa_blackwell")` from
`flashinfer-python==0.6.16.post2` against an FA4-lineage CuTe-DSL block-sparse
forward (`flash_attn.cute` BSA — the kernel family credited in your
`blk128/flash_fwd_sm100.py` header) on one GB200 (SM100). Two findings:
`vsa_blackwell` sustains ~60% of the FA4 kernel's throughput on identical
block masks, and its CPU-side `plan()` costs 1.3–2.7 ms per mask. Please take
a look at optimizing this backend. Results and repros below.

Context: we develop FastVideo/VSA (the workload integrated in #3250). VSA
masks are data-dependent and change every layer × step, so both findings sit
on its critical path.

**Setup (identical for both kernels):** batch 1, 12 heads, head_dim 128, bf16,
non-causal, kv_len = q_len; seq 4096–65536 × sparsity {0–90%}; seeded per-head
boolean block masks with identical selected-token sets for both kernels
(block-size note below); one sparse-aware FLOP formula
(`4·h·d·q_len·selected_blocks·block_tokens`); `triton.testing.do_bench`
(warmup=5, rep=20); `run()` timed alone, `plan()` timed separately. torch
2.12.0+cu130, CUDA 13.0, triton 3.7.0, driver 580.159.04. Outputs of both
kernels sit at the bf16 rounding floor vs an fp32 masked-SDPA reference in
every cell, so this is purely a throughput comparison.

**Block-size note.** The two kernels take sparsity masks at different
granularities: `vsa_blackwell` at 128×128 blocks (R=C=128), the FA4-lineage
kernel at 256-token tiles — on SM100 it rejects Q-side block sizes below 256
(verified), so 256 is its finest selection granularity. Masks are generated
once at VSA's 256-token tile granularity and expanded 2×2
(`repeat_interleave`) into the 128×128 block mask fed to `vsa_blackwell`, so
both kernels attend exactly the same tokens in every cell and the shared FLOP
count makes the TFLOP/s directly comparable.

**Finding 1 — flat ~950 TFLOP/s ceiling.** Achieved throughput is 887–961
TFLOP/s in every cell at seq ≥ 8192, essentially independent of sequence
length and sparsity — the FA4-lineage kernel sustains 1130–1620 on the same
masks (1.25–1.71× faster, median 1.60×):

| seq | sparsity | vsa_blackwell run / TFLOP/s | FA4-lineage CuTe BSA / TFLOP/s | time ratio | plan() ms |
|---|---|---|---|---|---|
| 16384 | dense | 1.810 ms / 911 | 1.186 ms / 1390 | 1.53× | 2.11 |
| 32768 | dense | 6.946 ms / 950 | 4.130 ms / 1597 | 1.68× | 2.22 |
| 65536 | dense | 28.030 ms / 941 | 17.695 ms / 1491 | 1.58× | 2.43 |
| 32768 | 90% (89.8) | 0.737 ms / 909 | 0.591 ms / 1133 | 1.25× | 2.21 |
| 49152 | 80% (80.2) | 3.076 ms / 955 | 1.955 ms / 1503 | 1.57× | 2.44 |
| 65536 | 90% (89.8) | 3.022 ms / 887 | 1.886 ms / 1421 | 1.60× | 2.24 |

Sparsity = fraction of 256-token KV tiles not attended per query tile;
keep = round((1−s)·NB) of NB tiles, self tile always kept — e.g. 90% at
seq 32768 is keep 13/128 = 89.8% exact (in parentheses). Per-cell keep
counts and `actual_sparsity` are in the harness JSON.

Below seq 8192 the comparison inverts: `vsa_blackwell` is faster there than
the FA4 wrapper stack.

**Finding 2 — `plan()` CPU cost on mask-per-call workloads.** With a boolean
`block_mask`, `plan()` costs 1.3–2.7 ms of CPU at seq ≥ 8192 (scales with
mask area; ~0.3 ms at 4096). VSA masks change every layer × step, so
mask-per-call deployments pay plan+run per call — at 32768/90% that is
2.2 ms plan + 0.74 ms run.

**Minimal repro** (single GPU, ~30 s, flashinfer-only deps):

```python
import time, torch
from triton.testing import do_bench
from flashinfer.sparse import BlockSparseAttentionWrapper

S, H, D, R = 32768, 12, 128, 128
NB = S // R
torch.manual_seed(42)
q = torch.randn(S, H, D, dtype=torch.bfloat16, device="cuda")
k, v = torch.randn_like(q), torch.randn_like(q)

g = torch.Generator("cpu").manual_seed(0)
keep = NB // 10  # ~90% sparse; keep = NB for dense
idx = torch.topk(torch.rand(H, NB, NB, generator=g), keep, -1).indices
mask = torch.zeros(H, NB, NB, dtype=torch.bool).scatter_(-1, idx, True).cuda()

w = BlockSparseAttentionWrapper(
    torch.empty(128 << 20, dtype=torch.uint8, device="cuda"),
    backend="vsa_blackwell")

def plan():
    w.plan(None, None, M=S, N=S, R=R, C=R, num_qo_heads=H, num_kv_heads=H,
           head_dim=D, block_mask=mask,
           q_data_type=torch.bfloat16, o_data_type=torch.bfloat16)

plan()  # first call includes JIT
torch.cuda.synchronize()
t0 = time.perf_counter(); plan(); torch.cuda.synchronize()
print(f"plan (steady): {(time.perf_counter() - t0) * 1e3:.2f} ms CPU")
ms = do_bench(lambda: w.run(q, k, v), warmup=5, rep=20)
flops = 4 * H * D * S * keep * R  # QK^T + PV over selected blocks only
print(f"run: {ms:.3f} ms   {flops / ms * 1e-9:.0f} TFLOP/s")
```

Expected on GB200: ~909 TFLOP/s run, ~2.2 ms plan. The FA4-lineage arm on the
identical mask: ~1133 TFLOP/s all-inclusive.

**Full grid harness** (both arms, identical masks/FLOPs, fp32 parity + bf16
floor per cell, plan/run split, markdown table + JSON output):
`bench_vsa_blackwell.py`:
https://gist.github.com/SolitaryThinker/90a1d1447929fc38dc509c1852e76532
The flashinfer arm needs only `flashinfer-python` + torch + triton; the FA4
arm runs when its deps (`fastvideo_kernel`, `quack`, `nvidia-cutlass-dsl`) are
importable and is skipped otherwise. `--quick` runs a 4-cell subset in ~1 min.

We can PR the harness into `flashinfer/benchmarks`, and can test candidate
branches on GB200. cc authors of #3250 / #4355 / #4397.
